ipykernel/subshell_manager.py

```diff
@@ -0,0 +1,283 @@
+
+        return subshell_id
+
+    def _delete_subshell(self, subshell_id: str) -> None:
+        """Delete subshell identified by subshell_id.
+
+        Raises key error if subshell_id not in cache.
+        """
+        assert current_thread().name == SHELL_CHANNEL_THREAD_NAME
+
+        with self._lock_cache:
+            subshell = self._cache.pop(subshell_id)
+
+        self._stop_subshell(subshell)
+
+    def _get_inproc_socket_address(self, name: str | None) -> str:
+        full_name = f"subshell-{name}" if name else "subshell"
+        return f"inproc://{full_name}"
+
+    def _get_shell_channel_socket(self, subshell_id: str | None) -> zmq.asyncio.Socket:
+        if subshell_id is None:
+            return self._parent_shell_channel_socket
+        with self._lock_cache:
+            return self._cache[subshell_id].shell_channel_socket
+
+    def _is_subshell(self, subshell_id: str | None) -> bool:
+        if subshell_id is None:
+            return True
+        with self._lock_cache:
+            return subshell_id in self._cache
+
+    async def _listen_for_subshell_reply(self, subshell_id: str | None) -> None:
+        """Listen for reply messages on specified subshell inproc socket and
+        resend to the client via the shell_socket.
+
+        Runs in the shell channel thread.
+        """
+        assert current_thread().name == SHELL_CHANNEL_THREAD_NAME
+
+        shell_channel_socket = self._get_shell_channel_socket(subshell_id)
+
+        try:
+            while True:
+                msg = await shell_channel_socket.recv_multipart(copy=False)
+                with self._lock_shell_socket:
+                    await self._shell_socket.send_multipart(msg)
+        except BaseException:
+            if not self._is_subshell(subshell_id):
+                # Subshell no longer exists so exit gracefully
+                return
+            raise
+
+    async def _process_control_request(
+        self, request: dict[str, t.Any], subshell_task: t.Any
+    ) -> dict[str, t.Any]:
+        """Process a control request message received on the control inproc
+        socket and return the reply.  Runs in the shell channel thread.
+        """
+        assert current_thread().name == SHELL_CHANNEL_THREAD_NAME
+
+        try:
+            type = request["type"]
+            reply: dict[str, t.Any] = {"status": "ok"}
+
+            if type == "create":
+                reply["subshell_id"] = await self._create_subshell(subshell_task)
+            elif type == "delete":
+                subshell_id = request["subshell_id"]
+                self._delete_subshell(subshell_id)
+            elif type == "list":
+                reply["subshell_id"] = self.list_subshell()
+            else:
+                msg = f"Unrecognised message type {type!r}"
+                raise RuntimeError(msg)
+        except BaseException as err:
+            reply = {
+                "status": "error",
+                "evalue": str(err),
```

Comment:
not for this PR:
 - Can this contain information that should not be send to the frontend ?
 - I don't believe so as the user is anyway the one that runs the kernel, but we had case where traceback would contain sensitive values. In other apps webapp this would be a likely issue (like revealing paths and filename of a server), but I don't think it's an issue here.